util/chunk/pool.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"container/list"
+	"math/rand"
+	"sync"
+
+	"github.com/pingcap/tidb/types"
+)
+
+// Pool is the column pool.
+type Pool struct {
+	varLenColPool   *colPool
+	fixLenColPool4  *colPool
+	fixLenColPool8  *colPool
+	fixLenColPool16 *colPool
+	fixLenColPool40 *colPool
+}
+
+// NewPool creates a new Pool.
+func NewPool(maxChunkSize int) *Pool {
+	numShards := 8
+	return &Pool{
+		varLenColPool:   newColPool(numShards, varElemLen),
+		fixLenColPool4:  newColPool(numShards, 4),
+		fixLenColPool8:  newColPool(numShards, 8),
+		fixLenColPool16: newColPool(numShards, 16),
+		fixLenColPool40: newColPool(numShards, 40),
+	}
+}
+
+// GetChunk gets a Chunk from the Pool.
+func (p *Pool) GetChunk(fields []*types.FieldType, cap int) *Chunk {
+	chk := new(Chunk)
+	chk.columns = make([]*column, 0, len(fields))
+	chk.capacity = cap
+
+	for _, f := range fields {
+		switch elemLen := getFixedLen(f); elemLen {
+		case varElemLen:
+			chk.columns = append(chk.columns, p.varLenColPool.get(cap))
+		case 4:
+			chk.columns = append(chk.columns, p.fixLenColPool4.get(cap))
+		case 8:
+			chk.columns = append(chk.columns, p.fixLenColPool8.get(cap))
+		case 16:
+			chk.columns = append(chk.columns, p.fixLenColPool16.get(cap))
+		case 40:
+			chk.columns = append(chk.columns, p.fixLenColPool40.get(cap))
+		}
+	}
+	return chk
+}
+
+// PutChunk puts a Chunk back to the Pool.
+func (p *Pool) PutChunk(fields []*types.FieldType, chk *Chunk) {
+	for i, f := range fields {
+		switch elemLen := getFixedLen(f); elemLen {
+		case varElemLen:
+			p.varLenColPool.put(chk.columns[i])
+		case 4:
+			p.fixLenColPool4.put(chk.columns[i])
+		case 8:
+			p.fixLenColPool8.put(chk.columns[i])
+		case 16:
+			p.fixLenColPool16.put(chk.columns[i])
+		case 40:
+			p.fixLenColPool40.put(chk.columns[i])
+		}
+	}
+}
+
+type colPool struct {
+	shards  []colPoolShard
+	elemLen int
+}
+
+func newColPool(numShards int, elemLen int) *colPool {
+	return &colPool{
+		shards:  make([]colPoolShard, numShards),
+		elemLen: elemLen,
+	}
+}
+
+func (cp *colPool) put(col *column) {
+	ordinal := rand.Int() % len(cp.shards)
```

Comment:
Does it matter that this still includes a global lock?

Reply:
```go
package main

import (
    "math/rand"
    "testing"
    "time"
)

func BenchmarkRandInt(b *testing.B) {
    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        rand.Int()
    }
}

func BenchmarkOwnRandInt(b *testing.B) {
    r := rand.New(rand.NewSource(time.Now().Unix()))

    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        r.Int()
    }
}
```

The benchmark result:
```
[jianzhang.zj:/tmp]
➜ go test -bench Benchmark -run xx -count 5 --benchmem
goos: darwin
goarch: amd64
BenchmarkRandInt-8              50000000                23.5 ns/op             0 B/op          0 allocs/op
BenchmarkRandInt-8              50000000                23.4 ns/op             0 B/op          0 allocs/op
BenchmarkRandInt-8              50000000                23.4 ns/op             0 B/op          0 allocs/op
BenchmarkRandInt-8              50000000                23.5 ns/op             0 B/op          0 allocs/op
BenchmarkRandInt-8              100000000               25.9 ns/op             0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.61 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.50 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.43 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.47 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.41 ns/op            0 B/op          0 allocs/op
PASS
ok      _/tmp   18.771s
```

Yes, it matters. I'll improve this.

Reply:
After investigating, I think we can calculate the hash value based on the memory address of the column, the hash function I used comes from https://stackoverflow.com/a/12996028/5303605, here is the code and benchmark result:
```go
func BenchmarkOwnRandInt(b *testing.B) {
    r := rand.New(rand.NewSource(time.Now().Unix()))

    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        _ = r.Int() % 8
    }
}

func BenchmarkOwnHash(b *testing.B) {
    val := 12
    pointer := &val

    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        x := *(*uint64)(unsafe.Pointer(&pointer))
        x = (x ^ (x >> 30)) * uint64(0xbf58476d1ce4e5b9)
        x = (x ^ (x >> 27)) * uint64(0x94d049bb133111eb)
        x = x ^ (x >> 31)
        _ = x % 8
    }
}
```

```
[jianzhang.zj:/tmp]
➜ go test -bench BenchmarkOwn --count 5 --benchmem
goos: darwin
goarch: amd64
BenchmarkOwnRandInt-8           200000000                7.49 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.50 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.51 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.45 ns/op            0 B/op          0 allocs/op
BenchmarkOwnRandInt-8           200000000                7.52 ns/op            0 B/op          0 allocs/op
BenchmarkOwnHash-8              2000000000               0.62 ns/op            0 B/op          0 allocs/op
BenchmarkOwnHash-8              2000000000               0.62 ns/op            0 B/op          0 allocs/op
BenchmarkOwnHash-8              2000000000               0.62 ns/op            0 B/op          0 allocs/op
BenchmarkOwnHash-8              2000000000               0.63 ns/op            0 B/op          0 allocs/op
BenchmarkOwnHash-8              2000000000               0.62 ns/op            0 B/op          0 allocs/op
PASS
ok      _/tmp   17.886s
```

Reply:
With this method, the columns with the same address are delivered to the same `colPoolShard`, we can add the deduplicate map in the `colPoolShard`. @XuHuaiyu